sound/soc/codecs/max98373-sdw.c

```diff
@@ -0,0 +1,970 @@
+		max98373->spkfb_slot = value & 0xF;
+	else
+		max98373->spkfb_slot = 2;
+
+	/* update interleave mode info */
+	if (device_property_read_bool(dev, "maxim,interleave_mode"))
+		max98373->interleave_mode = true;
+	else
+		max98373->interleave_mode = false;
+}
+
+/*     MAX98373 reset and initialization. */
+static int max98373_reset(struct device *dev)
+{
+	int ret, reg, count;
+	struct max98373_priv *max98373 = dev_get_drvdata(dev);
+
+	/* Software Reset */
+	ret = regmap_update_bits(max98373->regmap,
+				 MAX98373_R2000_SW_RESET,
+				 MAX98373_SOFT_RESET,
+				 MAX98373_SOFT_RESET);
+	if (ret) {
+		dev_err(dev, "Reset command failed. (ret:%d)\n", ret);
+		return ret;
+	}
+
+	for (count = 0; count < 3; count++) {
+		usleep_range(10000, 11000);
+		/* Software Reset Verification */
+		ret = regmap_read(max98373->regmap,
+				  MAX98373_R21FF_REV_ID, &reg);
+		if (!ret) {
+			dev_dbg(dev, "Reset completed (retry:%d)\n", count);
+			return 0;
+		}
+	}
+
+	dev_err(dev, "Reset failed. (ret:%d)\n", ret);
+	return ret;
+}
+
+/* Power management functions and structure */
+static int max98373_suspend(struct device *dev)
+{
+	struct max98373_priv *max98373 = dev_get_drvdata(dev);
+
+	regcache_cache_only(max98373->regmap, true);
+	regcache_mark_dirty(max98373->regmap);
+	return 0;
+}
+
+static int max98373_resume(struct device *dev)
+{
+	struct sdw_slave *slave = dev_to_sdw_dev(dev);
+	struct max98373_priv *max98373 = dev_get_drvdata(dev);
+	unsigned long time;
+
+	time = wait_for_completion_timeout(&slave->initialization_complete,
+					   msecs_to_jiffies(2000));
+	if (!time) {
+		dev_err(dev, "Initialization not complete, timed out\n");
+		return -ETIMEDOUT;
+	}
+
+	regcache_cache_only(max98373->regmap, false);
+	regcache_sync(max98373->regmap);
+
+	return 0;
+}
+
+static const struct dev_pm_ops max98373_pm = {
+	SET_SYSTEM_SLEEP_PM_OPS(max98373_suspend, max98373_resume)
+	SET_RUNTIME_PM_OPS(max98373_suspend, max98373_resume, NULL)
+};
+
+static int max98373_read_prop(struct sdw_slave *slave)
+{
+	struct sdw_slave_prop *prop = &slave->prop;
+	int nval, i, num_of_ports = 1;
+	u32 bit;
+	unsigned long addr;
+	struct sdw_dpn_prop *dpn;
+
+	/* BITMAP: 00001000  Dataport 3 is active */
+	prop->source_ports = BIT(3);
+	/* BITMAP: 00000010  Dataport 1 is active */
+	prop->sink_ports = BIT(1);
+	prop->paging_support = true;
+	prop->clk_stop_timeout = 20;
+
+	nval = hweight32(prop->source_ports);
+	num_of_ports += nval;
+	prop->src_dpn_prop = devm_kcalloc(&slave->dev, nval,
+					  sizeof(*prop->src_dpn_prop),
+					  GFP_KERNEL);
+	if (!prop->src_dpn_prop)
+		return -ENOMEM;
+
+	i = 0;
+	dpn = prop->src_dpn_prop;
+	addr = prop->source_ports;
+	for_each_set_bit(bit, &addr, 32) {
+		dpn[i].num = bit;
+		dpn[i].type = SDW_DPN_FULL;
+		dpn[i].simple_ch_prep_sm = true;
+		dpn[i].ch_prep_timeout = 10;
+		i++;
+	}
+
+	/* do this again for sink now */
+	nval = hweight32(prop->sink_ports);
+	num_of_ports += nval;
```

Comment:
the math seems incorrect for num_of_ports. If you have one source and one sink you'd end with num_of_ports = 3?

Reply:
OK. I was not sure the reason why num_of_ports is initialized by 1. Shall remove this, so that total num of ports become 2 not 3.

Reply:
Modified & re-submitted.. Please help review. Thanks